File: app/internal/doorpass/doorpass.go
// Package doorpass computes and verifies the rotating token shown on a
// member's digital pass. The QR encodes "<member_code>:<token>" — the
// member code tells the scanner whose secret to check the token against,
// the token itself proves the pass is live within the last WindowSeconds.
//
// This is the same scheme sketched in the prototype, made real: the token
// is HMAC-SHA256(door_secret, window) rather than random-looking canvas
// dots, and a server that holds the same secret can recompute and verify
// it without ever transmitting the secret over the wire on each scan.
package doorpass

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base32"
	"encoding/binary"
	"strings"
	"time"
)

// WindowSeconds is the rotation cadence. The frontend recomputes the QR at
// the same cadence purely from the secret it was handed at login — no
// network round trip needed to keep the pass fresh.
const WindowSeconds = 15

// ToleranceWindows is how many windows on either side of "now" the
// verifier accepts, absorbing clock drift and the second or two between a
// member's screen rendering a code and staff finishing the scan.
const ToleranceWindows = 1

// encoding produces upper-case, unpadded base32 — no ambiguous 0/O/1/I,
// reads cleanly on a small screen, matches the ABCDEFGH...23456789 style
// alphabet the prototype used for its display tokens.
var encoding = base32.StdEncoding.WithPadding(base32.NoPadding)

// CurrentWindow returns the rotation window index for a given time —
// unix-seconds divided by WindowSeconds, floored.
func CurrentWindow(t time.Time) int64 { return t.Unix() / WindowSeconds }

// Token computes the display token for one window: HMAC-SHA256(secret,
// window) truncated to 10 bytes and base32-encoded, which lands on exactly
// 16 characters (10 bytes = 80 bits = 16 groups of 5 bits) — grouped here
// as XXXX-XXXX-XXXX-XXXX for readability.
func Token(secret []byte, window int64) string {
	mac := hmac.New(sha256.New, secret)
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], uint64(window))
	mac.Write(buf[:])
	sum := mac.Sum(nil)[:10]

	raw := encoding.EncodeToString(sum) // 16 chars
	return raw[0:4] + "-" + raw[4:8] + "-" + raw[8:12] + "-" + raw[12:16]
}

// Verify checks a token against every window from now-ToleranceWindows to
// now+ToleranceWindows and reports the first match. Constant-time per
// comparison so a timing side-channel can't help an attacker narrow down
// the secret.
func Verify(secret []byte, token string, now time.Time) (ok bool, matchedWindow int64) {
	current := CurrentWindow(now)
	for d := -ToleranceWindows; d <= ToleranceWindows; d++ {
		w := current + int64(d)
		want := Token(secret, w)
		if subtle.ConstantTimeCompare([]byte(want), []byte(token)) == 1 {
			return true, w
		}
	}
	return false, 0
}

// Encode joins a member code and its current token into the single string
// the QR carries.
func Encode(memberCode, token string) string { return memberCode + ":" + token }

// Decode splits a scanned QR payload back into member code and token.
// Member codes never contain ':', so splitting on the last one is safe
// even if a code format changes shape later.
func Decode(payload string) (memberCode, token string, ok bool) {
	i := strings.LastIndex(payload, ":")
	if i < 0 {
		return "", "", false
	}
	return payload[:i], payload[i+1:], true
}

File: app/web/src/lib/doorpass.ts
// Mirrors internal/doorpass (Go) exactly: HMAC-SHA256(secret, window),
// truncated to 10 bytes, base32-encoded (16 chars), grouped as
// XXXX-XXXX-XXXX-XXXX. Computed entirely client-side from the secret
// fetched once at login — the pass keeps rotating with no network call.

export const WINDOW_SECONDS = 15

const BASE32_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567'

function base32Encode(bytes: Uint8Array): string {
  let bits = 0
  let value = 0
  let output = ''
  for (const b of bytes) {
    value = (value << 8) | b
    bits += 8
    while (bits >= 5) {
      output += BASE32_ALPHABET[(value >>> (bits - 5)) & 31]
      bits -= 5
    }
  }
  if (bits > 0) {
    output += BASE32_ALPHABET[(value << (5 - bits)) & 31]
  }
  return output
}

export function currentWindow(now: number = Date.now()): bigint {
  return BigInt(Math.floor(now / 1000 / WINDOW_SECONDS))
}

async function hmacSha256(keyBytes: Uint8Array, message: Uint8Array): Promise<Uint8Array> {
  // TS's DOM lib types BufferSource as backed by a plain ArrayBuffer, which
  // a generic Uint8Array<ArrayBufferLike> can't statically prove (it might
  // wrap a SharedArrayBuffer) — the cast is safe here since every caller in
  // this file constructs its arrays with `new Uint8Array(n)`.
  const key = await crypto.subtle.importKey(
    'raw', keyBytes as BufferSource, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign'],
  )
  const sig = await crypto.subtle.sign('HMAC', key, message as BufferSource)
  return new Uint8Array(sig)
}

function windowToBytes(window: bigint): Uint8Array {
  // Big-endian uint64, matching Go's binary.BigEndian.PutUint64.
  const buf = new Uint8Array(8)
  let w = window
  for (let i = 7; i >= 0; i--) {
    buf[i] = Number(w & 0xffn)
    w >>= 8n
  }
  return buf
}

/** Computes the display token for one rotation window. */
export async function token(secret: Uint8Array, window: bigint): Promise<string> {
  const mac = await hmacSha256(secret, windowToBytes(window))
  const raw = base32Encode(mac.slice(0, 10)) // 16 chars
  return `${raw.slice(0, 4)}-${raw.slice(4, 8)}-${raw.slice(8, 12)}-${raw.slice(12, 16)}`
}

/** Joins a member code and its current token into the QR payload. */
export function encode(memberCode: string, tok: string): string {
  return `${memberCode}:${tok}`
}

export function base64ToBytes(b64: string): Uint8Array {
  const bin = atob(b64)
  const bytes = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i)
  return bytes
}
